lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,119 @@
+<?php
+add_filter( 'rest_prepare_wp_navigation', 'gutenberg_hook_blocks_into_core_navigation_rest_response', 10, 3 );
+add_filter( 'block_core_navigation_render_inner_blocks', 'gutenberg_hook_blocks_into_core_navigation', 10, 3 );
+add_action( 'rest_insert_wp_navigation', 'gutenberg_update_ignore_hooked_blocks_meta_core_navigation', 10, 3 );
+
+/**
+ * Updates the post meta with the list of ignored hooked blocks when the navigation is created or updated via the REST API.
+ *
+ * @param WP_Post $post Post object.
+ */
+function gutenberg_update_ignore_hooked_blocks_meta_core_navigation( $post ) {
+	if ( ! isset( $post->ID ) ) {
+		return;
+	}
+
+	$anchor_block_type    = 'core/navigation';
+	$first_child_position = 'first_child';
+	$last_child_position  = 'last_child';
+
+	$hooked_blocks                 = get_hooked_blocks();
+	$hooked_block_types_last_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+	? $hooked_blocks[ $anchor_block_type ][ $last_child_position ]
+	: array();
+
+	$hooked_block_types_first_child = isset( $hooked_blocks[ $anchor_block_type ][ $first_child_position ] )
+	? $hooked_blocks[ $anchor_block_type ][ $first_child_position ]
+	: array();
+
+	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, null );
+	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, null );
+
+	$ignored_hooked_blocks = array_merge( $hooked_block_types_last_child, $hooked_block_types_first_child );
+
+	if ( ! empty( $ignored_hooked_blocks ) ) {
+		update_post_meta( $post->ID, '_wp_ignored_hooked_blocks', json_encode( $ignored_hooked_blocks ) );
+	}
```

Comment:
I think one thing that we need to accommodate here is that it's possible that the post already has some blocks stored in its `_wp_ignored_hooked_blocks` meta that we need to carry over.

The use case in which this could happen is something like:
- User enables plugin A, which inserts hooked block A into the Navigation block (e.g. Login/out block).
- User moves hooked block A around, saves Navigation menu. _Login/out block is added to `_wp_ignored_hooked_blocks`._
- User enabled plugin B, which inserts hooked block B (e.g. Mini Cart block).
- User makes some more changes to Navigation block, saves it. _Mini Cart block is added to `_wp_ignored_hooked_blocks`._

In the last step, we need to retain the Login/out block in `_wp_ignored_hooked_blocks`; otherwise, it'll be rendered twice (once because it's loaded from the Navigation menu in the DB that it's already been persisted two, and twice because Block Hooks are unaware that they need to continue to ignore it.)

Reply:
Fixed in https://github.com/WordPress/gutenberg/pull/57754/commits/248dfb024c042ed8c66af99b9e894bb71b41f9ae